torsiondrive/tests/test_extra_constraints.py

```diff
@@ -0,0 +1,220 @@
+"""
+Unit and regression test for the torsiondrive extra constraints feature
+"""
+
+import pytest
+from torsiondrive.extra_constraints import make_constraints_dict, check_conflict_constraits, \
+    build_geometric_constraint_string, build_terachem_constraint_string
+
+
+def test_make_constraints_dict():
+    '''
+    Unit test for torsiondrive.qm_engine.make_constraints_dict() function
+    '''
+    constraits_string = '''
+    $freeze
+    xyz 1-3,6-7
+    distance 1 5
+    $set
+    distance 1 4 1.0
+    angle 4 1 2 30.0
+    dihedral 1 2 3 4 60.0
+    '''
+    constraints_dict = make_constraints_dict(constraits_string)
+    # test the freeze constraints
+    spec_list = constraints_dict['freeze']
+    assert len(spec_list) == 2
+    assert spec_list[0]['type'] == 'xyz'
+    assert spec_list[0]['indices'] == [0, 1, 2, 5, 6]
+    assert spec_list[1]['type'] == 'distance'
+    assert spec_list[1]['indices'] == [0, 4]
+    # test the set constraints
+    spec_list = constraints_dict['set']
+    assert len(spec_list) == 3
+    assert spec_list[0]['type'] == 'distance'
+    assert spec_list[0]['indices'] == [0, 3]
+    assert spec_list[0]['value'] == 1.0
+    assert spec_list[1]['type'] == 'angle'
+    assert spec_list[1]['indices'] == [3, 0, 1]
+    assert spec_list[1]['value'] == 30.0
+    assert spec_list[2]['type'] == 'dihedral'
+    assert spec_list[2]['indices'] == [0, 1, 2 ,3]
+    assert spec_list[2]['value'] == 60.0
+    # test error handling for wrong inputs
+    # scan not supported
+    wrong_constraits_string = '''
+    $scan
+    distance 1 2 1.0 2.0 10
+    '''
+    with pytest.raises(ValueError):
```

Comment:
I like that you specifically look for `ValueError` here. One way to be even safer with the test that raises an exception is to set `pytest.raises(ValueError, match="some part of the error message")`. I've had tests that weren't working as intended because, while they were raising the exception type I was looking for, it was coming up in the wrong part of the code.

Reply:
I agree this could be helpful. Will add the `match=ErrorCode` feature later when we have a more consolidated error code.